### PR TITLE
TER-94 export from TD TPT to ABS

### PR DESCRIPTION
**Pull request by martinjunger**
Jira: TER-94 (export), TER-98 (import)

Before asking for review make sure that:

- [x] In the case of **modified import queries**, I checked that the [CustomQuery component](https://github.com/keboola/app-custom-query-manager) is updated.

---

- load test files to ABS: `composer loadABS`
- update `.env`:
  - fill `ABS_*` variables
  - change `STORAGE_TYPE` variable to `ABS`
  - ABS import a export:
    - run test `\Tests\Keboola\Db\ImportExportFunctional\Teradata\ExportTest::testExportSimple`
    - run test `\Tests\Keboola\Db\ImportExportFunctional\Teradata\ToFinal\FullImportTest::testFullImportWithDataSet`

### CODE REVIEW

**Comment by jirkasemmler** [comment]
na prvni pohled to vypada dobre, nezkoumal jsem to detailne, chtelo by to prvne videt zelene testy

File: src/Backend/Teradata/Helper/StorageS3Helper.php [hunk 1]
```diff
@@ -7,13 +7,8 @@
 use Keboola\Db\ImportExport\Backend\Helper\BackendHelper as BaseHelper;
 use Keboola\Db\ImportExport\Storage\S3\SourceFile;
 
-final class BackendHelper extends BaseHelper
+final class StorageS3Helper extends BaseHelper
 {
-    public static function quoteValue(string $value): string
```
Comment (jirkasemmler) [inline src/Backend/Teradata/Helper/StorageS3Helper.php:12]: proc?

File: src/Backend/Teradata/ToStage/FromABSTPTAdapter.php [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+use Keboola\Db\ImportExport\Backend\Teradata\Helper\StorageABSHelper;
+use Keboola\Db\ImportExport\Backend\Teradata\Helper\StorageS3Helper;
+use Keboola\Db\ImportExport\Backend\Teradata\TeradataImportOptions;
+use Keboola\Db\ImportExport\Backend\Teradata\ToFinalTable\SqlBuilder;
+use Keboola\Db\ImportExport\Backend\Teradata\ToStage\Exception\FailedTPTLoadException;
+use Keboola\Db\ImportExport\Backend\Teradata\ToStage\Exception\NoMoreRoomInTDException;
+use Keboola\Db\ImportExport\ImportOptions;
+use Keboola\Db\ImportExport\ImportOptionsInterface;
+use Keboola\Db\ImportExport\Storage;
+use Keboola\FileStorage\Path\RelativePath;
```
Comment (jirkasemmler) [inline src/Backend/Teradata/ToStage/FromABSTPTAdapter.php:18]: tyhlety dva usy nepotrebujes

File: src/Storage/ABS/SourceFile.php [hunk 3]
```diff
@@ -44,13 +46,15 @@ public function __construct(
...
-        ?array $primaryKeysNames = null
+        ?array $primaryKeysNames = null,
+        ?string $blobMasterKey = null
     ) {
         parent::__construct($container, $filePath, $sasToken, $accountName);
         $this->isSliced = $isSliced;
         $this->csvOptions = $csvOptions;
         $this->columnsNames = $columnsNames;
         $this->primaryKeysNames = $primaryKeysNames;
+        $this->blobMasterKey = $blobMasterKey;
```
Comment (jirkasemmler) [inline src/Storage/ABS/SourceFile.php:57]: co to je?

File: tests/functional/Teradata/ExportTest.php [hunk 4]
```diff
@@ -244,6 +244,65 @@ public function testExportSimple(): void
         $this->assertContains($this->getExportDir() . '/ts_test/ts_testmanifest', array_values($files));
     }
 
+    public function testExportSimpleAbs(): void
```
Comment (jirkasemmler) [inline tests/functional/Teradata/ExportTest.php:247]: proc to potrebujeme, v cem se to lisi?

**Review by jirkasemmler — CHANGES_REQUESTED** [review]
cool, potrebujeme poresit ten manifest. jen se mi nezda ten helper, nebo jsem ho aspon nepochopil

File: tests/functional/Teradata/ExportTest.php (hunk 4 above)
Comment (martinjunger) [reply tests/functional/Teradata/ExportTest.php:247]: ted uz to nepotrebujeme, to byl prvni pokus pro Export na ABS (jeste bez importu)

File: src/Storage/ABS/SourceFile.php (hunk 3 above)
Comment (martinjunger) [reply src/Storage/ABS/SourceFile.php:57]: to je potreba v TPT, tak to tam musim nejak dostat.

File: src/Backend/Teradata/Helper/QuoteHelper.php [hunk 5]
```diff
@@ -0,0 +1,13 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Keboola\Db\ImportExport\Backend\Teradata\Helper;
+
+final class QuoteHelper
```
[the remark this answers is not in the document]
Comment (martinjunger) [reply src/Backend/Teradata/Helper/QuoteHelper.php:7]: protoze jsem prejmenoval `...Backend/Teradata/Helper/BackendHelper.php` na `...ckend/Teradata/Helper/StorageS3Helper.php`, tak jsem oddelil pryc.

Podobne to ma Snowflake `src/Backend/Snowflake/Helper/QuoteHelper.php`

File: tests/functional/ImportExportBaseTest.php [hunk 6]
```diff
@@ -14,6 +14,7 @@
 abstract class ImportExportBaseTest extends TestCase
 {
     public const TABLE_OUT_CSV_2COLS = 'out_csv_2Cols';
+    public const TABLE_OUT_CSV_2COLS_NO_INSERT = 'out_csv_2Cols_no_insert';
```
Comment (jirkasemmler) [inline tests/functional/ImportExportBaseTest.php:17]: na co potrebujeme toto?

Comment (jirkasemmler) [reply tests/functional/ImportExportBaseTest.php:17]: a k tomu tests/functional/Teradata/TeradataBaseTestCase.php ? protoze nikde to nemuzu najit, ze by se takova tabulka nakonec pouzila

File: tests/functional/Teradata/ExportTest.php (hunk 4 above)
Comment (martinjunger) [reply tests/functional/Teradata/ExportTest.php]: odstraneno v commitu [remove temporary export test](https://github.com/keboola/php-db-import-export/pull/165/commits/24e22d032f02bf854c5b42d6a0e21a562c5bf0f9)

Comment (jirkasemmler) [reply tests/functional/Teradata/ExportTest.php]: ok, smazu to

File: tests/functional/ImportExportBaseTest.php (hunk 6 above)
Comment (martinjunger) [reply tests/functional/ImportExportBaseTest.php:17]: to je ta z toho odstranenyho docasnyho testu ([tady](https://github.com/keboola/php-db-import-export/pull/165/commits/24e22d032f02bf854c5b42d6a0e21a562c5bf0f9)) - muze se smazat

File: src/Backend/Teradata/Helper/QuoteHelper.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (jirkasemmler) [reply src/Backend/Teradata/Helper/QuoteHelper.php:7]: jo, ale nikde se to nepouziva, nebo jo ?

[the remark this answers is not in the document]
Comment (martinjunger) [reply src/Backend/Teradata/Helper/QuoteHelper.php:7]: hmm, to mas pravdu, asi to byla nejak puvodni copy&paste ze Snflk - tak to smaz

**Comment by jirkasemmler** [comment]
@martinjunger za me je to takto ok. snad ty testy tedka projdou. Prosim jen to nejak po-fixupuj, at pak neni v masteru takovej mrdnik commitu :)

**Comment by martinjunger** [comment]
> @martinjunger za me je to takto ok. snad ty testy tedka projdou. Prosim jen to nejak po-fixupuj, at pak neni v masteru takovej mrdnik commitu :)

Trochu jsem to squasnul.

**Review by jirkasemmler — APPROVED** [review]

**Pull request merged**